### ISSUE TITLE
can't call method "do" on an undefined value at /usr/local/share/perl5/Ora2pg.pm

### ISSUE THREAD

**Issue by hebronie** [darold/ora2pg#542]
When I run the command "ora2pb" or other variant like "ora2pg -u system -w password -o /home/oracle/output.sql",  I get the following

can't call method "do" on an undefined value at /usr/local/share/perl5/Ora2Pg.pm line 5221
Issueing rollback() due to DESTROY without explicit disconnect() of DBD::db handle (DESCRIPTION=(ADDRESS=(HOST=servername.com)(PROTOCOL=tcp)(PORT=1521))(CONNECT_DATA=(SID=dbname))) at /usr/local/share/perl5/Ora2Pg.pm line 5220

What can I do? I am running this on the oracle server. The PG is on another server. same network. I tried other user besides "system". same error.

**Comment by darold** [comment]
What is your PG_DSN value? What version of Ora2Pg are you using?

**Comment by hebronie** [comment]
PG_DSN    dbi:Pg:dbname=misa_misb;host=server.domain.com;port=5432

Pra2Pg - ver 18.2

Thanks.

Aaron

**Comment by darold** [comment]
Sorry I have misunderstood your issue. I think you can not just connect to your Oracle database. Can you connect to your database using sqlplus from this server using:
```
sqlplus system/password@servername.com/db911
```

**Comment by hebronie** [comment]
Yes, I can connect.

**Comment by darold** [comment]
Then my understanding was right, the error reported:
```
can't call method "do" on an undefined value at /usr/local/share/perl5/Ora2Pg.pm line 5221
```
is related to the PostgreSQL connection. Can you use the -d option to ora2pg command and post all reported message?

**Comment by hebronie** [comment]
[oracle@oracledb ~]$ ora2pg -d
    Ora2Pg version: 18.2
    Trying to connect to database: dbi:Oracle:host=servername.com;sid=DB911;port=1521
    Isolation level: SET TRANSACTION ISOLATION LEVEL SERIALIZABLE
    Looking forward functions declaration in schema U911OPS.
    Retrieving table information...
    Retrieving table information...
    Retrieving table information...
    Retrieving views information...
    Retrieving users/roles/grants information...
    Retrieving sequences information...
    Retrieving triggers information...
    Retrieving functions information...
    Retrieving procedures information...
    Retrieving tablespaces information...
    ...........................
    Retrieving user defined types information...
    -- Generated by Ora2Pg, the Oracle database Schema converter, version 18.2
    -- Copyright 2000-2017 Gilles DAROLD. All rights reserved.
    -- DATASOURCE: dbi:Oracle:host=servername.com;sid=DB911;port=1521
    SET client_encoding TO 'UTF8';
    \set ON_ERROR_STOP ON
    SET search_path = u911ops;
    Trying to connect to database: dbi:Oracle:host=servername.com;sid=DB911;port=1521
    Isolation level: SET TRANSACTION ISOLATION LEVEL SERIALIZABLE
    Retrieving partitions information...
    Can't call method "do" on an undefined value at /usr/local/share/perl5/Ora2Pg.pm
    line 5221.
    Issuing rollback() due to DESTROY without explicit disconnect() of 
    DBD::Oracle::db handle (DESCRIPTION=(ADDRESS=(HOST=localhost)(PROTOCOL=tcp)
    (PORT=1521))(CONNECT_DATA=(SID=DB911))) at /usr/local/share/perl5/Ora2Pg.pm line
    5220.
    [oracle@oracledb ~]$

**Comment by hebronie** [comment]
TYPE    TABLE COPY VIEW GRANT SEQUENCE TRIGGER FUNCTION PROCEDURE TABLESPACE

I confirmed that PG database name and schema  is available

**Comment by darold** [comment]
You can not use data export (COPY) together with other export type. Commit 45fe50d add a note about that in the documentation.

**Comment by hebronie** [comment]
Thanks for all your help. What do you suggest I remove or Add?

**Comment by darold** [comment]
I recommend you to use `ora2pg --init_project mig_prj_name` then after `cd mig_prj_name/` edit the config/ora2pg.conf file to set the ORACLE_HOME, ORACLE_DSN, SCHEMA and PG_DSN configuration directives. Test your settings using:
```
ora2pg -c config/ora2pg.conf -t SHOW_VERSION
```
When the connection to Oracle works, just execute script `export_all.sh`, it will do all the job and place all output files in dedicated directories.

**Comment by hebronie** [comment]
It looks like its working. Now, how do I get all this into PG. Oracle and PG server are different server node. Also, can Ora2pg export over network from Oracle into PG database. Please state the step like you did last one. It was extremely useful.

**Comment by hebronie** [comment]
Note: after export_all.sh script, output files are all in place

**Comment by hebronie** [comment]
Hello. After following the last instruction, It ask me to run "ora2pg -t COPY -o data.sql  -b ./data -c ./config/ora2pg.conf" It gives me the following:
[================>]  15/15 tables (100%) end of scanning.
DBD::Pg::db do failed: ERROR: relation "event_cross_reference" does not exist at /usr/local/share/perl5/Ora2Pg.pm line 5277.
FATAL: ERROR: relation "event_cross_reference" does not exist
Aborting export...

**Comment by darold** [comment]
Ora2Pg doesn't import DDL into PostgreSQL, you have to done this manually using psql and see what should be adapted/fixed. Once you have validated all DDL you can create all tables and run the command to export data. The DDL files are all under the `schema/` subdirectory.

**Comment by hebronie** [comment]
Thanks Darold. I have done the the DDL in the schema subdirectory. After which I started the ora2pg -t COPY..... But I am getting errors. I reran the command again and got same thing. I ran the command after these error, it seems the "out of memory" error comes on the same spot; table/row.

$ ora2pg -t COPY -o data.sql -b ./data -c ./config/ora2pg.conf
[========================>] 15/15 tables (100.0%) end of scanning.                       
[=====>                   ]  560037/2347650 rows (23.9%) Table TABLE_A_1 (20001 recs/sec)
[>                        ]   560037/90689408 total rows (0.6%) - (28 sec., avg: 20001 recs/sec).   
[========================>] 597575/1 rows (59757500.0%) Table TABLE_A_2 (29878 recs/sec)      
[>                        ]  1157612/90689408 total rows (1.3%) - (48 sec., avg: 24116 recs/sec).       
[========================>] 331896/1 rows (33189600.0%) Table TABLE_A_3 (30172 recs/sec)      
[>                        ]  1489508/90689408 total rows (1.6%) - (59 sec., avg: 25245 recs/sec).       
[=====>                   ]  877484/3631922 rows (24.2%) Table TABLE_B_1 (2309 recs/sec)            
[>                        ]  2366992/90689408 total rows (2.6%) - (439 sec., avg: 5391 recs/sec). 
Out of memory!            ]  5979574/90689408 rows (6.6%) on total estimated data (439 sec., avg: 13620 recs/sec)
Issuing rollback() due to DESTROY without explicit disconnect() of DBD::Oracle::db handle (DESCRIPTION=(ADDRESS=(HOST=localhost)(PROTOCOL=tcp)(PORT=1001))(CONNECT_DATA=(SID=dbname))) at /usr/local/lib64/perl5/DBD/Oracle.pm line 346.

**Comment by hebronie** [comment]
Thanks. It works very well.

**Comment by hebronie** [comment]
The first schema migration was success full. But while doing the second one, I noticed the following error below. Is there anyway I can fix it or just skip it.

$ ora2pg -t COPY -o data.sql -b ./data -c ./config/ora2pg.conf
[========================>] 15/15 tables (100.0%) end of scanning.                       
[=====>                   ]  560037/2347650 rows (23.9%) Table TABLE_A_1 (20001 recs/sec)
[>                        ]   560037/90689408 total rows (0.6%) - (28 sec., avg: 20001 recs/sec).   
[========================>] 597575/1 rows (59757500.0%) Table TABLE_A_2 (29878 recs/sec)      
[>                        ]  1157612/90689408 total rows (1.3%) - (48 sec., avg: 24116 recs/sec).       
[========================>] 331896/1 rows (33189600.0%) Table TABLE_A_3 (30172 recs/sec)      
[>                        ]  1489508/90689408 total rows (1.6%) - (59 sec., avg: 25245 recs/sec).       
[=====>                   ]  877484/3631922 rows (24.2%) Table TABLE_B_1 (2309 recs/sec)            
[>                        ]  2366992/90689408 total rows (2.6%) - (439 sec., avg: 5391 recs/sec). 

DBD::Pg::db do failed: ERROR: relation "jauary2010" does not exists at /usr/local/share/perl5/Ora2Pg.pm line 12746.ec)
FATAL: ERROR: relation "january2010" does not exist
Abort export...

**Comment by darold** [comment]
I guess that you have imported the tables in a different schema in PostgreSQL. YOu have two possibilities, the first one is to set the schema at user connection (alter role ... set search_path to ...) or enable EXPORT_SCHEMA and PG_SCHEMA if the schema name have been changed.

**Comment by blabblab** [comment]
Hello,
 I am also having some issues in the migration. I m using latest version(18) of Ora2pg and trying to do direct import from Oracle(11g) to Fujitsu Enterprise PostgreSQL (9.5).
 I am not receiving any error but I can not see any objects in PG database.
Below is the output of ora2pg command and ora2pg conf file
-------------
-bash-4.1$ ora2pg
[========================>] 7/7 tables (100.0%) end of scanning.
[========================>] 7/7 tables (100.0%) end of table export.
[========================>] 0/0 packages (100.0%) end of output.
[========================>] 1/1 views (100.0%) end of output.
[========================>] 3/3 sequences (100.0%) end of output.
[========================>] 1/1 triggers (100.0%) end of output.
[========================>] 0/0 functions (100.0%) end of functions export.
[========================>] 2/2 procedures (100.0%) end of procedures export.
[========================>] 0/0 types (100.0%) end of output.
[========================>] 0/0 partitions (100.0%) end of output.
------------
Below is my conf file.
--------------------------------
_ORACLE_HOME     /u01/app/oracle/product/11.2.0/db_1
ORACLE_DSN      dbi:Oracle:host=ol6-112.localdomain;sid=DB11G;port=1521
ORACLE_USER     system
ORACLE_PWD      oracle
SCHEMA          HR
#EXPORT_SCHEMA   1
COMPILE_SCHEMA  1
TYPE          TABLE  PACKAGE  VIEW GRANT SEQUENCE TRIGGER FUNCTION PROCEDURE TABLESPACE 
 TYPE PARTITION
DISABLE_COMMENT 1
DROP_FKEY       1
OUTPUT          HR_output.sql
FILE_PER_FKEYS  1
FILE_PER_CONSTRAINT  1
FKEY_DEFERRABLE 1
DEFER_FKEY      1
DATA_TYPE      DATE:timestamp,LONG:text,LONG RAW:bytea,CLOB:text,NCLOB:text,BLOB:bytea,BFILE:bytea,RAW:bytea,ROWID:oid,FLOAT:double precision,DEC:decimal,DECIMAL:decimal,DOUBLE PRECISION:do
uble precision,INT:numeric,INTEGER:numeric,REAL:real,SMALLINT:smallint,BINARY_FLOAT:double precision,BINARY_DOUBLE:double precision,TIMESTAMP:timestamp,XMLTYPE:xml,BINARY_INTEGER:integer,PL
S_INTEGER:integer,TIMESTAMP WITH TIME ZONE:timestamp with time zone,TIMESTAMP WITH LOCAL TIME ZONE:timestamp with time zone ,number:numeric,varchar2:varchar
#PG_NUMERIC_TYPE 1
PG_SCHEMA       hr
PG_DSN          dbi:Pg:dbname=db11g;host=ol6-112.localdomain;port=27500
PG_USER         fsep
PG_PWD          ******_

Also, when I am trying to execute 'CONSTRAINTS_HR_output.sql', I am receiving blow error 

_ERROR:  constraints on permanent tables may reference only permanent tables (11988)
STATEMENT:  ALTER TABLE countries ADD CONSTRAINT countr_reg_fk FOREIGN KEY (region_id) REFERENCES regions(region_id) ON DELETE NO ACTION DEFERRABLE INITIALLY DEFERRED;_**

**Comment by SaanSingh** [comment]
Thanks for your reply.

I tried it and it worked for me. 

But I am still facing below issue -
1- Constraint 
ERROR:  constraints on permanent tables may reference only permanent tables (11988)
STATEMENT:  ALTER TABLE countries ADD CONSTRAINT countr_reg_fk FOREIGN KEY (region_id) REFERENCES regions(region_id) ON DELETE NO ACTION DEFERRABLE INITIALLY DEFERRED;

2- I am just testing it for HR schema, my 'regions' table is not owned by 'hr' rather by superuser.

3- Many indexes are also missing on the table, mainly unique indexes. Please see the example below.

== ORACLE:
SQL> select INDEX_NAME,STATUS from DBA_INDEXES where TABLE_NAME='COUNTRIES';

INDEX_NAME                     STATUS
------------------------------ --------
COUNTRY_C_ID_PK                VALID
COUNTRIES_PK                   VALID
COUNTRY_C_ID_PK                VALID

==PG
db11g=# \d countries
               Table "hr.countries"
    Column    |         Type          | Modifiers
--------------+-----------------------+-----------
 country_id   | character(2)          | not null
 country_name | character varying(40) |
 region_id    | bigint                |
Indexes:
    "countries_pkey" PRIMARY KEY, btree (country_id)
Note that Indexes issue is mostly for all the tables.


Thanks in advance.

**Comment by darold** [comment]
Can you post the definition of table `regions` in PostgreSQL? It may have been declared as UNLOGGED table.

**Comment by SaanSingh** [comment]
Actually, all the tables in postgreSQL are showing 'Unlogged Table'. Below are the desc of regions and employees tables.

db11g=# \d regions
           Unlogged table "hr.regions"
   Column    |         Type          | Modifiers
-------------+-----------------------+-----------
 region_id   | bigint               | not null
 region_name | character varying(25) |
Indexes:
    "regions_pkey" PRIMARY KEY, btree (region_id)

db11g=#
db11g=#
db11g=# \d employees
              Unlogged table "hr.employees"
     Column     |            Type             | Modifiers
----------------+-----------------------------+-----------
 employee_id    | integer                     | not null
 first_name     | character varying(20)       |
 last_name      | character varying(25)       | not null
 email             | character varying(25)       | not null
 phone_number   | character varying(20)       |
 hire_date      | timestamp without time zone | not null
 job_id         | character varying(10)       | not null
 salary         | numeric(8,2)                |
 commission_pct | numeric(2,2)                |
 manager_id     | integer                     |
 department_id  | smallint                    |
Indexes:
    "employees_pkey" PRIMARY KEY, btree (employee_id)
    "employees_email_key" UNIQUE CONSTRAINT, btree (email)
    "employees_department_id" btree (department_id)
    "employees_job_id" btree (job_id)
    "employees_last_name_first_name" btree (last_name, first_name)
    "employees_manager_id" btree (manager_id)
Check constraints:
    "emp_salary_min" CHECK (salary > 0::numeric)
Foreign-key constraints:
    "emp_dept_fk" FOREIGN KEY (department_id) REFERENCES departments(department_id)
    "emp_job_fk" FOREIGN KEY (job_id) REFERENCES jobs(job_id)
    "emp_manager_fk" FOREIGN KEY (manager_id) REFERENCES employees(employee_id)
Referenced by:
    TABLE "departments" CONSTRAINT "dept_mgr_fk" FOREIGN KEY (manager_id) REFERENCES employees(employee_id)
    TABLE "employees" CONSTRAINT "emp_manager_fk" FOREIGN KEY (manager_id) REFERENCES employees(employee_id)
    TABLE "job_history" CONSTRAINT "jhist_emp_fk" FOREIGN KEY (employee_id) REFERENCES employees(employee_id)
Triggers:
    update_job_history AFTER UPDATE ON employees FOR EACH ROW EXECUTE PROCEDURE trigger_fct_update_job_history()

**Comment by darold** [comment]
I guess that this is because your Oracle tables have been declared with LOGGING = NO. Have a look to:
```
SELECT  TABLE_NAME, LOGGING FROM ALL_TABLES
```

**Comment by SaanSingh** [comment]
Thanks for the reply.

I remove it manually from DDL before migrating it to PG.
But my concern is , all other tables DDL are also having Logging but not having any issue in PG except regions.

What is the best approach to handle this issue?
Thanks

**Comment by darold** [comment]
If there is no issue with other tables this is because there is no other FK referencing a column in an unlogged table. The best approach is what you have done, manually removed UNLOGGED keyword from the DDL script. You can also change this attribute in Oracle if you want.

**Comment by SaanSingh** [comment]
Thanks for your response.
May I ask one more question- is there any way to do the migration assessment for Oracle .sql file( take export and then import in .sql file) or it can be done only after connecting to the Oracle database ?

Simply, my Oracle DB is not accessible to Ora2pg ( client is not ready to install it) and I am having sql(DDL) file of binary dump.

Thanks in advance.

**Comment by darold** [comment]
Yes you can have a SQL text file that you can give to ora2pg with the -i option then choose the right export type (QUERY, FUNCTION, PROCEDURE or PACKAGE) you want with the -t option then add the --estimate_cost option.

Sorry for the response delay.
